### ISSUE TITLE
Building Packages with Fable Compatible F# Dependencies 

### ISSUE THREAD

**Issue by cloudRoutine** [fable-compiler/Fable#278]
I started thinking about this when I was wondering what the best way to enable projects like [Gjallarhorn](http://reedcopsey.github.io/Gjallarhorn/) to publish Fable compatible packages.

There seem to be a few possible approaches 
1. Nuget packages that include the source files from the subset of the project that will be included with fable and fable manifest file. Similar to how [Cargo works with Crates](http://doc.crates.io/guide.html)
2. NPM packages of standard compiled js
3. Fables plugins to adapt the library
4. Git dependencies with paket that use a Fable manifest in root to know what to build  

The main features that are important to be preserve are the design time tooling and type checking that come from the dll references in the Fable project being built. 

One advantage of using a method that use the F# source to build everything together is it could enable the fable compiler to do dead code elimination to reduce the size of the final Fable package.

**Comment by alfonsogarciacaro** [comment]
My preferred solution would be some variance/evolution of the way shown for [the fable helper sample](https://www.npmjs.com/package/fable-helpers-sample), i.e.: distributing the F# dll and the compiled JS code (either through npm or nuget).

**Plugins** are a bit complicated, and I would recommend them only when the whole library cannot be compiled to JS or we are only interested in some part of it (e.g. FSharp.Core, NUnit).

I wouldn't recommend **distributing the F# source to compile it to JS** either (though this will always be a last resort as the consumer can just reference the github repo with Paket) as this will slow/complicate the whole compilation and some tweaks may be necessary to compile to JS which should be responsible of the lib author.

> The main features that are important to be preserve are the design time tooling and type checking that come from the dll references in the Fable project being built.

This is possible if the consumer project just references the lib dll.

> One advantage of using a method that use the F# source to build everything together is it could enable the fable compiler to do dead code elimination to reduce the size of the final Fable package.

Fable doesn't do dead code elimination by itself. However, when referencing a type from another file it always used [ES6 member imports](https://developer.mozilla.org/en/docs/web/javascript/reference/statements/import). This, combined with tools like Webpack 2 and uglify-js, allows for [tree shaking](http://www.2ality.com/2015/12/webpack-tree-shaking.html). The condition to allow this feature would be that the lib author compiles to JS using the `--module es2015` fable compiler option.

In conclusion, it's already possible to distribute F# packages and reference them with fable using the `--refs` option. I think what remains to be decided is:

1. _Should we always distribute JS code using ES6 modules?_ This would allow tree shaking. The problems would be: a) cannot be used directly in node or the browser (like UMD), and b) only Webpack 2 (still in beta, I think) supports ES6 imports.
2. _Should we add some "manifest" file next to `dll` so Fable can automatically recognise where's the compiled JS code?_ This will avoid the need for the consumer project to use the `--refs` argument in `fableconfig.json`.

**Comment by cloudRoutine** [comment]
>  The condition to allow this feature would be that the lib author compiles to JS using the --module es2015 fable compiler option.

This information and possibly some other metadata could also be stored in the manifest

**Comment by cloudRoutine** [comment]
Once  a format has been decided on a FAKE helper module could be created to simplify and streamline the process of creating nuget or npm fable packages

**Comment by ReedCopsey** [comment]
Just brainstorming... would it be possible for fable to add the compiled js as Metadata into the assembly?  could potentially allow single file f#+js libs...

**Comment by alfonsogarciacaro** [comment]
@ReedCopsey, this could be an interesting idea. Actually we considered something similar but just to add the fable manifest data into the assembly. However this may make things a bit more complicated and for now probably it's better to stick to human readable files that are easily editable.

I gave it a quick thought and maybe the manifest could be name something like `fablemanifest.json` (I know, very imaginative) and contain the following fields (names are tentative):
* `assemblyName`: The name of the F# `dll` in case there are more assemblies in the same file. Perhaps the manifest itself could include info for different `dlls`?
* `js` / `jsCode` / `jsLocation` / `jsDir`: Indicates a dir relative to the current location where the JS code is to be found. It could be possible to point to an npm package instead (so the dll would be distributed through nuget and the js through npm).
* `module`: The module system used to compile the code to JS.
* `fableCoreVersion`: The version of `fable-core` used when compiling the code to JS.

**Comment by cloudRoutine** [comment]
> It could be possible to point to an npm package instead (so the dll would be distributed through nuget and the js through npm
That's a really cool idea!

**Comment by alfonsogarciacaro** [comment]
Fable 0.7 includes a new way to distribute packages using `.dll` and `.fablemap`, still through npm though. Enrico is also working to integrate npm packages with the new fsproj format. Let's wait a bit to see the response from the community and see if it makes sense to unify distribution of .NET and JS.
